homeassistant/components/mqtt/config_flow.py

```diff
@@ -337,38 +450,151 @@ def validate_field(
         errors[field] = error
 
 
+@callback
+def _check_conditions(
+    platform_field: PlatformField, component: dict[str, Any] | None = None
+) -> bool:
+    """Only include field if one of conditions match, or no conditions are set."""
+    if platform_field.conditions is None or component is None:
+        return True
+    return any(
+        all(component.get(key) == value for key, value in condition.items())
+        for condition in platform_field.conditions
+    )
+
+
+@callback
+def calculate_merged_config(
+    merged_user_input: dict[str, Any],
+    data_schema_fields: dict[str, PlatformField],
+    component_data: dict[str, Any],
+) -> dict[str, Any]:
+    """Calculate merged config."""
+    base_schema_fields = {
+        key
+        for key, platform_field in data_schema_fields.items()
+        if _check_conditions(platform_field, component_data)
+    } - set(merged_user_input)
+    return {
+        key: value
+        for key, value in component_data.items()
+        if key not in base_schema_fields
+    } | merged_user_input
+
+
 @callback
 def validate_user_input(
     user_input: dict[str, Any],
     data_schema_fields: dict[str, PlatformField],
     errors: dict[str, str],
-) -> None:
+    component_data: dict[str, Any] | None,
+    config_validator: Callable[
+        [dict[str, Any], dict[str, str], list[str] | None], dict[str, str]
+    ]
+    | None = None,
+) -> dict[str, Any]:
     """Validate user input."""
-    for field, value in user_input.items():
+    # Merge sections
+    reset_fields: list[str] = []
+    merged_user_input: dict[str, Any] = {}
+    for key, value in user_input.items():
+        # Omit empty lists that are not allowed to be empty
```

Comment:
I don't understand this comment. The comment suggests we're omitting empty list, but the code is about updating existing dicts?

Reply:
There is no way users can clear the field. If the selector was filled with options before, we want to be able to clear the list.

Reply:
Will remove the comment.